### PR TITLE
Fixes #116: Reuse OkHttpClient Instance

### LINKED ISSUE #116: Reuse OkHttpClient instance
It is a best practice to create an `OkHttpClient` instance only once and then reuse it.
If new configuration is needed, use `client.newBuilder` on it [to share resources like thread pools](https://square.github.io/okhttp/3.x/okhttp/okhttp3/OkHttpClient.html) (read the paragraph about customizing the client).

- [IconRequestModelLoader:119](https://github.com/d4rken/reddit-android-appstore/blob/815be90708cd3370e3abb8d91b039b7672f48721/app/src/main/java/subreddit/android/appstore/util/ui/glide/IconRequestModelLoader.java#L119)
- [LiveWikiRepository:31](https://github.com/d4rken/reddit-android-appstore/blob/a40705eb6645b5c7bcdfdbde0a149a2b79c73267/app/src/main/java/subreddit/android/appstore/backend/reddit/wiki/LiveWikiRepository.java#L31)
- [LiveWikiRepository:54](https://github.com/d4rken/reddit-android-appstore/blob/a40705eb6645b5c7bcdfdbde0a149a2b79c73267/app/src/main/java/subreddit/android/appstore/backend/reddit/wiki/LiveWikiRepository.java#L54)
- [GPlayScraper:26](https://github.com/d4rken/reddit-android-appstore/blob/815be90708cd3370e3abb8d91b039b7672f48721/app/src/main/java/subreddit/android/appstore/backend/scrapers/gplay/GPlayScraper.java#L26)
- [LiveGithubRepository:30](https://github.com/d4rken/reddit-android-appstore/blob/31500e5ba5e91d702fc707eab771dd11e7c04947/app/src/main/java/subreddit/android/appstore/backend/github/LiveGithubRepository.java#L30)
- [TokenRepository:55](https://github.com/d4rken/reddit-android-appstore/blob/815be90708cd3370e3abb8d91b039b7672f48721/app/src/main/java/subreddit/android/appstore/backend/reddit/TokenRepository.java#L55)

### PR DESCRIPTION
**Pull request by bobheadxi**
As pointed out in #116, OkHttpClient instances should be reused. With the `@ApplicationScope` [annotation](https://github.com/d4rken/reddit-android-appstore/blob/dev/app/src/main/java/subreddit/android/appstore/util/dagger/ApplicationScope.java), I think `OkHttpClient` etc instances will be automatically reused by Dagger, not certain on that though

- All `OkHttpClient`, `Retrofit` and `Gson` dependencies are now provided by Dagger
- Removed all other occurrences of `OkHttpClient` creation
- Minor refactor I missed in #100

### CODE REVIEW

File: app/src/main/java/subreddit/android/appstore/backend/reddit/wiki/LiveWikiRepository.java [hunk 1]
```diff
@@ -135,21 +113,22 @@ private String saveAuthString(String authString) {
     }
 
     interface WikiApi {
-        @GET("r/Android/wiki/page")
+        // BASEURL overides retrofit.baseUrl()
+        @GET(BASEURL + "r/Android/wiki/page")
```
Comment (d4rken) [inline app/src/main/java/subreddit/android/appstore/backend/reddit/wiki/LiveWikiRepository.java]: Not a fan of hardcoding the baseurl. Why not set it through the Retrofit instance?

File: app/src/main/java/subreddit/android/appstore/backend/HttpModule.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    @Provides
+    @ApplicationScope
+    Retrofit provideRetrofit(OkHttpClient client) {
+        return new Retrofit.Builder()
+                .client(client)
+                .addConverterFactory(GsonConverterFactory.create())
+                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
+                .baseUrl("https://api.github.com/")
```
Comment (d4rken) [inline app/src/main/java/subreddit/android/appstore/backend/HttpModule.java]: We specify the base url here, but also in the annotation in the interfaces?

File: app/src/main/java/subreddit/android/appstore/backend/reddit/TokenRepository.java [hunk 3]
```diff
@@ -105,8 +83,9 @@ void storeToken(@NonNull Token token) {
     }
 
     interface TokenApi {
+        // BASEURL overides retrofit.baseUrl()
```
Comment (d4rken) [inline app/src/main/java/subreddit/android/appstore/backend/reddit/TokenRepository.java]: I don't think that's how it works. AFAIK we now have `https://api.github.com/https://oauth.reddit.com/api/v1/access_token`?

File: app/src/main/java/subreddit/android/appstore/backend/HttpModule.java (hunk 2 above)
Comment (bobheadxi) [inline app/src/main/java/subreddit/android/appstore/backend/HttpModule.java]: It appears that it is not possible to instantiate Retrofit without specifying a baseUrl

File: app/src/main/java/subreddit/android/appstore/backend/reddit/TokenRepository.java (hunk 3 above)
Comment (bobheadxi) [inline app/src/main/java/subreddit/android/appstore/backend/reddit/TokenRepository.java]: As of Retrofit2, [including full URL will supersede the baseUrl](https://stackoverflow.com/questions/34842390/how-to-setup-retrofit-with-no-baseurl)

Comment (d4rken) [inline app/src/main/java/subreddit/android/appstore/backend/reddit/TokenRepository.java]: TIL!

File: app/src/main/java/subreddit/android/appstore/backend/HttpModule.java (hunk 2 above)
Comment (d4rken) [inline app/src/main/java/subreddit/android/appstore/backend/HttpModule.java]: I think we should inject the okhttp client and the endpoint url, but not the retrofit instance, or inject the `Retrofit.Builder`, depending on which approach is user to mock in unit tests.

Each network component have it's own retrofit instance and should decide itself how the retrofit instance is build, in our simple case this currently just affects the baseurl, but in more complex cases, different endpoints would also need different converter and adapter factories.

Comment (bobheadxi) [inline app/src/main/java/subreddit/android/appstore/backend/HttpModule.java]: Good point about unit tests, just a thought but would named dependencies work for this? ie providing a different Retrofit instance for LiveWikiRepository and LiveGithubRepository

Comment (d4rken) [inline app/src/main/java/subreddit/android/appstore/backend/HttpModule.java]: Yes though at this point we could also directly inject the api instance as the repos don't need access to the Retrofit instance . We'd need named deps to provide different base urls though.

Comment (bobheadxi) [inline app/src/main/java/subreddit/android/appstore/backend/HttpModule.java]: @d4rken I opted to inject the APIs in 2178a16

Not sure if what I've done with the classes is considered best practice - for example, for `LiveWikiRepository`, I moved the API and response types into `Wiki`, so the API is accessed with `Wiki.Api` and response types with `Wiki.SomeResponseType`

Also unsure if I should continue to use @ApplicationScope for the API instances

**Comment by d4rken** [comment]
> Not sure if what I've done with the classes is considered best practice - for example, for LiveWikiRepository, I moved the API and response types into a Wiki class, so the API is accessed with Wiki.Api and response types with Wiki.SomeResponseType

That's quite common, but I think it should be done the other way around, e.g we have a `WikiApi` , the token is a `WikiApi.Token` because the token is specific to this api. Similar to "package by feature".

> Also unsure if I should continue to use @ApplicationScope for the API instances

The annotation causes the provided instance to last as long as the corresponding dagger component  stays alive, with `ApplicationScope` this usually refers to the dagger component referenced from `App`. Because the api instances never change, we can keep this annotated as we don't need different instances during the lifetime of the app.

**Comment by bobheadxi** [comment]
@d4rken if this looks good I'll merge?

**Comment by d4rken** [comment]
lgtm

**Pull request merged**